### DISCUSSION TITLE
Extension settings changed programatically don't show in UI

### DISCUSSION THREAD

**Discussion by sebastian-correa** [Ulauncher/Ulauncher#897] [Extensions and Themes]
### Problem
I am trying to implement a reset to default "button" as a select preference. The entry in `manifest.json` looks like this:
```
{
      "id": "reset_to_default",
      "type": "select",
      "name": "Reset preferences to default",
      "default_value": "-",
      "options": [
        {
          "value": "-",
          "text": "-"
        },
        {
          "value": "Reset",
          "text": "Reset"
        }
      ]
}
```
And the `PreferencesUpdateListener` looks like this:
```
DEFAULT_PREFERENCES = {"kw": "rae", "max_suggested_items": 10, "max_shown_definitions": 10, "reset_to_default": "-"}
class PreferencesUpdateListener(EventListener):
    def on_event(self, event, extension):
        if event.id == "reset_to_default" and event.new_value == 'Reset':
            for id_, val in DEFAULT_PREFERENCES.items():
                extension.preferences[id_] = val
```
When I toggle the Reset button to reset and _save_, the extension's behaviour changes but the UI doesn't reflect that.

Source can be found [here](https://github.com/sebastian-correa/ulauncher-rae-dle).

### Steps to reproduce
1. Define `PreferencesUpdateListener` with an `on_event` method that updates a preference with a predefined value.
2. Before changing values in the UI, test extension to verify it complies with the behaviour reflected in the UI.
3. Trigger the `on_event` method (in my case, by toggling a button in the preferences menu).
4. Verify the extension now behaves as expected, but that the UI shows the old values.

See [attached video for illustration](https://user-images.githubusercontent.com/43179146/117025537-5dfdf980-acd1-11eb-9f8e-ec011c876100.mp4).

### Current behavior
`PreferencesUpdateListener` resets the `extension.preferences` property to its default value and the behaviour changes accordingly. That is, when I use the extension, the new set values are in effect.

However, the extensions settings UI shows the old values.

### Expected behavior
When I change the `extension.preferences` property in a `PreferencesUpdateListener`, both the behaviour and the values in the settings UI should reflect these new values.

### Environment
**Ulauncher version**: 5.11.0
**Linux distro**: Pop_OS! 20.10
**Desktop environment**: Gnome 3.38.3

### Logs
Not sure whether they add much information, but [here](https://gist.github.com/sebastian-correa/6f6a0061b6d6d962ca7c277cc6f43dd1) they are. They correspond to the process shown in the [video](https://user-images.githubusercontent.com/43179146/117025537-5dfdf980-acd1-11eb-9f8e-ec011c876100.mp4).

### Closing remarks
This might not be a bug. I suspect I have to tell the extension to update the UI once the values are set in the code. I did not find a way to do this, and this template suits the issue the best. Feel free to change the tags.

Thanks!

**Comment by friday** [comment]
We don't really support this. The users should be in control of the user settings. It seems like you found a way to override that. If we directly supported a "Reset" function like your is would be different, but I'm not sure we want to do that.

**Reply by sebastian-correa** [reply]
Hey, thanks for your patience!! Busy end/start of year for me.

Fair enough. My intent was to have a safety for a user that breaks something. If they did, they could just revert the settings to the default.

Maybe having the `default_value` key auto show up in the settings would be a nice middle ground, where I don't have to remember to include it (and change it if the default changes for some reason) in the `description` key but wouldn't signify a big change in functionality from your side. It would also help to cohesion, because currently different extensions could show a field's default values in different ways.

Options that come to mind are:
1. A line below the `description`.
2. Right next to the `name`.
3. As (phantom) text to the right inside the box with the option.

Cheers!!
